src/quo/components/wallet/swap_input/view.cljs

```diff
@@ -0,0 +1,118 @@
+(ns quo.components.wallet.swap-input.view
+  (:require [quo.components.avatars.token-avatar.view :as token-avatar]
+            [quo.components.buttons.button.view :as buttons]
+            [quo.components.dividers.divider-line.view :as divider-line]
+            [quo.components.markdown.text :as text]
+            [quo.components.tags.network-tags.view :as network-tag]
+            [quo.components.wallet.approval-label.schema :as approval-label.schema]
+            [quo.components.wallet.approval-label.view :as approval-label]
+            [quo.components.wallet.swap-input.style :as style]
+            [quo.foundations.colors :as colors]
+            quo.theme
+            [react-native.core :as rn]
+            [schema.core :as schema]))
+
+(def ?schema
+  [:=>
+   [:catn
+    [:props
+     [:map {:closed true}
+      [:type {:optional true} [:maybe [:enum :pay :receive]]]
+      [:status {:optional true} [:maybe [:enum :default :disabled :loading]]]
+      [:token {:optional true} [:maybe :string]]
+      [:value {:optional true} [:maybe :string]]
+      [:default-value {:optional true} [:maybe :string]]
+      [:currency-symbol {:optional true} [:maybe :string]]
+      [:fait-value {:optional true} [:maybe :string]]
+      [:show-approval-label? {:optional true} [:maybe :boolean]]
+      [:error? {:optional true} [:maybe :boolean]]
+      [:approval-label-props {:optional true} [:maybe approval-label.schema/?schema]]
+      [:network-tag-props {:optional true} [:maybe :map]]
+      [:on-change-text {:optional true} [:maybe fn?]]
+      [:enable-swap? {:optional true} [:maybe :boolean]]
+      [:on-swap-press {:optional true} [:maybe fn?]]
+      [:on-token-press {:optional true} [:maybe fn?]]
+      [:on-max-press {:optional true} [:maybe fn?]]
+      [:customization-color {:optional true} [:maybe :schema.common/customization-color]]
+      [:container-style {:optional true} [:maybe :map]]]]]
+   :any])
+
+(defn view-internal
+  [{:keys [type status token value fait-value show-approval-label? error? network-tag-props
+           approval-label-props default-value enable-swap?
+           currency-symbol on-change-text
+           container-style on-swap-press on-token-press on-max-press]}]
+  (let [theme             (quo.theme/use-theme)
+        pay?              (= type :pay)
+        disabled?         (= status :disabled)
+        loading?          (= status :loading)
+        controlled-input? (some? value)
+        input-ref         (rn/use-ref-atom nil)
```

Comment:
@ajayesivan _Not really a suggestion for change_, but rather a technique we could use more often: we can extract stateful parts and isolate them inside custom hooks. Maybe here it's not the best example, but if we consider this technique more often, we might be able to pull off some reusable hooks in the codebase, who knows, patterns could emerge.

Example also with @smohamedjavid and @ulisesmac suggestions.

```diff
modified   src/quo/components/wallet/swap_input/view.cljs
@@ -1,5 +1,6 @@
 (ns quo.components.wallet.swap-input.view
-  (:require [quo.components.avatars.token-avatar.view :as token-avatar]
+  (:require [oops.core :as oops]
+            [quo.components.avatars.token-avatar.view :as token-avatar]
             [quo.components.buttons.button.view :as buttons]
             [quo.components.dividers.divider-line.view :as divider-line]
             [quo.components.markdown.text :as text]
@@ -37,20 +38,30 @@
       [:container-style {:optional true} [:maybe :map]]]]]
    :any])
 
+(defn- use-input
+  []
+  (let [input-ref     (rn/use-ref-atom nil)
+        set-input-ref (rn/use-callback #(reset! input-ref %) [])
+        focus-input   (rn/use-callback
+                       (fn []
+                         (some-> @input-ref
+                                 (oops/ocall :focus)))
+                       [input-ref])]
+    {:input-ref     input-ref
+     :set-input-ref set-input-ref
+     :focus-input   focus-input}))
+
 (defn view-internal
   [{:keys [type status token value fait-value show-approval-label? error? network-tag-props
            approval-label-props default-value enable-swap?
            currency-symbol on-change-text
            container-style on-swap-press on-token-press on-max-press]}]
-  (let [theme             (quo.theme/use-theme)
-        pay?              (= type :pay)
-        disabled?         (= status :disabled)
-        loading?          (= status :loading)
-        controlled-input? (some? value)
-        input-ref         (rn/use-ref-atom nil)
-        set-input-ref     (rn/use-callback (fn [ref] (reset! input-ref ref)) [])
-        focus-input       (rn/use-callback (fn [] (.focus ^js @input-ref))
-                                           [input-ref])]
+  (let [theme                               (quo.theme/use-theme)
+        pay?                                (= type :pay)
+        disabled?                           (= status :disabled)
+        loading?                            (= status :loading)
+        controlled-input?                   (some? value)
+        {:keys [set-input-ref focus-input]} (use-input)]
```

Reply:
is a common practice to mix multiple hooks to create a custom one in react?

Since they manage state, I'm a bit worried they hide too much the complexity, or that we end up creating abstractions for the state that hide important details and we end-up suing helper funcitons to modify that state.

Maybe not too bad actually, but it's a significantly different way to write components

Reply:
In my experience with React it is @ulisesmac, quite common. The idea is that with hooks the API is more functional instead of the old OO way, so it's easier to decouple. It's almost copy & paste and then parameterize the arguments, so it's a cheap refactor available to us. I would certainly not use this all the time. This here is maybe too simple of an example, but the basic idea holds for more complex scenarios.

I remember reading in React docs that one of the goals of hooks is to give us a way to extract, refactor and reuse stateful operations. Now that the code is really into hooks, it's an opportunity to explore.